File: c++/211.cpp
// 211. Add and Search Word - Data structure design
// Design a data structure that supports the following two operations:
//
// void addWord(word)
// bool search(word)
// search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.
//
// Example:
//
// addWord("bad")
// addWord("dad")
// addWord("mad")
// search("pad") -> false
// search("bad") -> true
// search(".ad") -> true
// search("b..") -> true
// Note:
// You may assume that all words are consist of lowercase letters a-z.

// Time Limit Exceeded
class WordDictionary {
public:
  std::vector<string> words;
  /** Initialize your data structure here. */
  WordDictionary() {
  }
  
  /** Adds a word into the data structure. */
  void addWord(string word) {
      this->words.push_back(word);
  }

  /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
  bool search(string word) {
    int length = words.size();
    for (int i = 0; i < length; i++) {
      if(std::regex_match(this->words[i], std::regex(word))) {
        return true;
      }
    }
    return false;
  }
};

// prefix tree
// The regex in this problem only includes '.', which means that there is no other special characters.
// 52 ms, faster than 96.54%
class WordDictionary {
public:
  struct Node {
  public:
    Node *child[26];
    bool isWord;
    Node(): isWord(false) {
      for (auto &c : child) c = NULL;
    }
  };

  WordDictionary() {
    root = new Node();
  }

  void addWord(string word) {
    Node *p = root;
    int i = 0;
    for (auto &c : word) {
      i = c - 'a';
      if (!p->child[i]) p->child[i] = new Node();
        p = p->child[i];
      }
      p->isWord = true;
  }
  bool search(string word) {
    return searchWord(word, root, 0);
  }

  bool searchWord(string &word, Node *p, int i) {
    if (i == word.size()) return p->isWord;
    if (word[i] == '.') {
      for (auto &c : p->child) {
        if (c && searchWord(word, c, i + 1)) return true;
      }
      return false;
    }
    else {
      return p->child[word[i] - 'a'] && searchWord(word, p->child[word[i] - 'a'], i + 1);
    }
  }

private:
  Node *root;
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */

File: go/211.go
// 211. Design Add and Search Words Data Structure

// Design a data structure that supports adding new words and finding if a string matches any previously added string.

// Implement the WordDictionary class:

// WordDictionary() Initializes the object.
// void addWord(word) Adds word to the data structure, it can be matched later.
// bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.
 

// Example:

// Input
// ["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
// [[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
// Output
// [null,null,null,null,false,true,true,true]

// Explanation
// WordDictionary wordDictionary = new WordDictionary();
// wordDictionary.addWord("bad");
// wordDictionary.addWord("dad");
// wordDictionary.addWord("mad");
// wordDictionary.search("pad"); // return False
// wordDictionary.search("bad"); // return True
// wordDictionary.search(".ad"); // return True
// wordDictionary.search("b.."); // return True
 

// Constraints:

// 1 <= word.length <= 25
// word in addWord consists of lowercase English letters.
// word in search consist of '.' or lowercase English letters.
// There will be at most 2 dots in word for search queries.
// At most 104 calls will be made to addWord and search.

// Runtime 323 ms Beats 76.19%
// Memory 67.02 MB Beats 34.92%
type WordDictionary struct {
    NextLayer []*WordDictionary
    HasEnd bool
}


func Constructor() WordDictionary {
    return WordDictionary{
        NextLayer: make([]*WordDictionary, 26),
    }
}


func (this *WordDictionary) AddWord(word string)  {
    cur := this
    for i, char := range word {
        index := char - rune('a')
        if cur.NextLayer[index] == nil {
            cur.NextLayer[index] = &WordDictionary{
                NextLayer: make([]*WordDictionary, 26),
            }
        }
        cur = cur.NextLayer[index]
        if i == len(word) - 1 {
            cur.HasEnd = true
        }
    }
}


func (this *WordDictionary) Search(word string) bool {
    cur := this
    for i, char := range word {
        if char == rune('.') {
            hasResult := false
            for j := 0; j < 26; j++ {
                if cur.NextLayer[j] != nil {
                    hasResult = hasResult || cur.NextLayer[j].Search(word[i + 1:])
                    if hasResult == true {
                        return true
                    }
                }
            }
            return false
        } else {
            index := char - rune('a')
            if cur.NextLayer[index] == nil {
                return false
            }
            cur = cur.NextLayer[index]
        }
    }
    return cur.HasEnd == true
}


/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */